source/simulator/solver_schemes.cc

```diff
@@ -778,11 +795,34 @@ namespace aspect
 
     double relative_residual = std::numeric_limits<double>::max();
     nonlinear_iteration = 0;
+
+    // Store the current location of the particles, so that we can reset
+    // the particles to it in each nonlinear iteration.
+    unsigned int n_locally_owned_particles = 0;
+    std::vector< Point<dim> > local_particle_positions(n_locally_owned_particles);
+    if (particle_world.get() != nullptr)
+      {
+        n_locally_owned_particles = particle_world->get_particle_handler().n_locally_owned_particles();
+        local_particle_positions.resize(n_locally_owned_particles);
+        // Get the current particle positions
+        particle_world->get_particle_handler().get_particle_positions(local_particle_positions);
+      }
+
     do
       {
         const double relative_temperature_residual =
           assemble_and_solve_temperature(nonlinear_iteration == 0, &initial_temperature_residual);
 
+        // Reset particle positions and advect them with the current velocity
+        if (particle_world.get() != nullptr)
+          {
+            // use 'false' to overwrite the current position of the particles
+            // TODO from manual: The new set of points defined by the vector has to be
+            // sufficiently close to the original one to ensure that the sort_particles_into_subdomains_and_cells()
+            // function manages to find the new cells in which the particles belong.
+            particle_world->get_particle_handler().set_particle_positions(local_particle_positions, false);
```

Comment:
same here, I had no idea ... the `ParticleHandler` has gotten a life on its own since it was added to deal.II (thanks to Luca Heltai and Bruno Blais :-))

Reply:
However there is a problem here. What happens to the particle properties? They are also updated when `advance_timestep` is called. I think we may need to store a full copy of the `ParticleHandler` and reset that every iteration.